src/lib/blit.c

```diff
@@ -711,6 +711,448 @@ sextant_blit(ncplane* nc, int linesize, const void* data, int leny, int lenx,
+    uint32_t l0 = generalerp(rsum0, gsum0, bsum0, insum);
+    uint32_t l1 = generalerp(rsum1, gsum1, bsum1, outsum);
+    uint32_t totaldiff = 0;
+    for(unsigned mask = 0 ; mask < 8 ; ++mask){
+      unsigned r, g, b;
+      if(glyph & (1u << mask)){
+        ncchannel_rgb8(l0, &r, &g, &b);
+      }else{
+        ncchannel_rgb8(l1, &r, &g, &b);
+      }
+      uint32_t rdiff = rgb_diff(ncpixel_r(rgbas[mask]), ncpixel_g(rgbas[mask]),
+                                ncpixel_b(rgbas[mask]), r, g, b);
+      totaldiff += rdiff;
+    }
+    if(totaldiff < mindiff){
+      mindiff = totaldiff;
+      best = glyph;
+      ncchannels_set_fchannel(channels, l0);
+      ncchannels_set_bchannel(channels, l1);
+    }
+    if(totaldiff == 0){ // can't beat that!
+      break;
+    }
+  }
+  assert(best >= 0 && best < 128);
+  if(blendcolors){
+    ncchannels_set_fg_alpha(channels, NCALPHA_BLEND);
+    ncchannels_set_bg_alpha(channels, NCALPHA_BLEND);
+  }
+  return octant_egcs[best];
+}
+
+static const char*
+oct_trans_check(nccell* c, const uint32_t rgbas[8], unsigned blendcolors,
+                uint32_t transcolor, unsigned nointerpolate){
+  unsigned transstring = 0;
+  unsigned r = 0, g = 0, b = 0;
+  unsigned div = 0;
+  for(unsigned mask = 0 ; mask < 8 ; ++mask){
+    if(rgba_trans_p(rgbas[mask], transcolor)){
+      transstring |= (1u << mask);
+    }else if(!nointerpolate || !div){
+      r += ncpixel_r(rgbas[mask]);
+      g += ncpixel_g(rgbas[mask]);
+      b += ncpixel_b(rgbas[mask]);
+      ++div;
+    }
+  }
+  if(transstring == 0){ // there was no transparency
+    return NULL;
+  }
+  nccell_set_bg_alpha(c, NCALPHA_TRANSPARENT);
+  // there were some transparent pixels. since they get priority, the foreground
+  // is just a general lerp across non-transparent pixels.
+  const char* egc = octant_egcs[transstring ^ 0xff];
+  nccell_set_bg_alpha(c, NCALPHA_TRANSPARENT);
+//fprintf(stderr, "transtring: %u egc: %s\n", transtring, egc);
+  if(*egc == ' '){ // entirely transparent
+    nccell_set_fg_alpha(c, NCALPHA_TRANSPARENT);
+    return "";
+  }else{ // partially transparent, thus div >= 1
+//fprintf(stderr, "div: %u r: %u g: %u b: %u\n", div, r, g, b);
+    cell_set_fchannel(c, generalerp(r, g, b, div));
+    if(blendcolors){
+      nccell_set_fg_alpha(c, NCALPHA_BLEND);
+    }
+    cell_set_blitquadrants(c, !(transstring & 5u), !(transstring & 10u),
+                              !(transstring & 20u), !(transstring & 40u));
+  }
+//fprintf(stderr, "OCT-BQ: 0x%x\n", cell_blittedquadrants(c));
+  return egc;
+}
+
+// octant blitter. maps 4x2 to each cell. since we only have two colors at
+// our disposal (foreground and background), we lose some fidelity.
+static inline int
+octant_blit(ncplane* nc, int linesize, const void* data, int leny, int lenx,
+            const blitterargs* bargs){
+  const unsigned nointerpolate = bargs->flags & NCVISUAL_OPTION_NOINTERPOLATE;
+  const bool blendcolors = bargs->flags & NCVISUAL_OPTION_BLEND;
```

Comment:
there's a lot of cut-and-paste going on here. i understand that, but one thing i was hoping to do was to come back and parameterize a general geometric character blitter. if you think that's beyond your abilities / the scope of this pr, that's fine, but it's something i'd like to do sooner rather than later.

maybe i ought do a parameterization, and then fold your patch into it? unsure.

reviewing this is definitely....dicey.

i think at a minimum i'd like to see this share the braille solver code. it seems like they ought be the exact same logic, just with two distinct sets of EGCs, no? i think that's a requirement for getting this in, unless you can show me why they'd be different solvers. what do you think?

Reply:
I currently replicated the search-over-everything logic from the sextant solver. I can switch to using the Braille solver logic. The EGCs for octants are not ordered regularly, so there will need to be a lookup table. If you're fine with using a lookup table for Braille as well then I can combine the solvers.

Reply:
yeah let's use the braille code plus a lookup table. if it's easiest, we can just use a lookup table for the braille, too, and pass them in as parameters.

Reply:
Here's the generic blitter.

Reply:
beautiful. i'll try to look at this today, but it might not be until tomorrow. thanks a lot for your work here. fwiw, i think this will likely be the largest contribution to the notcurses c core by someone who is not me.